### PR TITLE
Pipeline API additions: merge(), distinct() and simpler co-aggregation methods

### PR DESCRIPTION
**Pull request by mtopolnik**

### CODE REVIEW

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/BatchStage.java [hunk 1]
```diff
@@ -74,6 +80,45 @@
...
+     * @return the newly attached stage
+     */
+    @Nonnull
+    default BatchStage<T> distinct() {
+        return distinctBy(wholeItem());
+    }
+
+    /**
+     * Attaches to this stage a stage that emits all the items from this
+     * stage as well as all the items from the supplied stage. The other
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/BatchStage.java]: I found the wording here slightly too elaborate, it could be expressed in simpler terms perhaps?

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/GroupAggregateBuilder1.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+ * Offers a step-by-step fluent API to build a pipeline stage that
+ * co-groups and aggregates the data from several input stages. To obtain
+ * it, call {@link StageWithGrouping#aggregateBuilder()} on one of the
+ * stages to co-group and refer to that method's Javadoc for further
+ * details.
+ *
+ * @param <T0> type of the stream-0 item
+ * @param <K> type of the grouping key
+ */
+public class GroupAggregateBuilder1<T0, K> {
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/GroupAggregateBuilder1.java]: should go through nullable annotations here as well

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/WindowGroupAggregateBuilder1.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+ * the builder) by calling {@link #tag0()}.
+ * <p>
+ * This object is mainly intended to build a co-aggregation of four or more
+ * contributing stages. For up to three stages, prefer the direct {@code
+ * stage.aggregateN(...)} calls because they offer more static type safety.
+ *
+ * @param <T0> the type of the stream-0 item
+ * @param <K> type of the key
+ */
+public class WindowGroupAggregateBuilder1<T0, K> {
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/WindowGroupAggregateBuilder1.java]: missing nullable annotations again

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/StageWithWindow.java [hunk 4]
```diff
@@ -63,6 +71,39 @@
             @Nonnull DistributedFunction<? super T, ? extends K> keyFn
     );
 
+    /**
+     * Javadoc pending.
```
Comment (viliam-durina) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/StageWithWindow.java]: Javadoc pending. Few of this more below.
